### PR TITLE
Fix: Ignore cards reviewed before settable to future values

### LINKED ISSUE #5284: Ignore cards reviewed before settable to dates in the future
Could you try these steps:

1. Set the ignore before to a future date
2. Create a new card
3. Review it 
4. Check the card info (This is where it appears as you describe)
5. Manually change parameters slightly
6. Check it again (This is where it appears as I describe)

I'm guessing there's actually a weird inconsistency here because having an ignore cards before date in the future from when you are reviewing isn't intended behaviour :thinking: I think we should make setting ignore before to a future date impossible.

_Originally posted by @Luc-Mcgrady in https://github.com/ankitects/anki/issues/5277#issuecomment-5205131415_

### PR DESCRIPTION
**Pull request by Luc-Mcgrady**
closes #5284

> (self quote) As I see it. I can't think of a use for setting the ignore_reviews_before value in the future. It also seems to result in undefined behaviour so I think we should prevent it from being set like that.

### CODE REVIEW

File: rslib/src/deckconfig/update.rs [hunk 1]
```diff
@@ -173,6 +174,10 @@ impl Collection {
 
         // add/update provided configs
         for conf in &mut req.configs {
+            require!(
```
Comment (fernandolins) [inline rslib/src/deckconfig/update.rs]: What happens if a user already has a future date saved?
Since the check runs over every config on each save, would they be blocked from saving any deck-option change until they manually reset that date, even when editing an unrelated field?

File: ts/routes/deck-options/AdvancedOptions.svelte [hunk 2]
```diff
@@ -303,7 +303,10 @@ License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
             </SpinBoxFloatRow>
 
             <Item>
-                <DateInput bind:date={$config.ignoreRevlogsBeforeDate}>
+                <DateInput
+                    bind:date={$config.ignoreRevlogsBeforeDate}
+                    max={new Date().toISOString().split("T")[0]}
```
Comment (fernandolins) [inline ts/routes/deck-options/AdvancedOptions.svelte]: Should this use UTC or the local date? As Anki runs locally, I think it should match the backend and use the local date. What do you think?

Comment (Luc-Mcgrady) [reply ts/routes/deck-options/AdvancedOptions.svelte]: Actually if we are being strict about this we need to consider the rollover of the cards too. If it is 1am with the rollover at 4am then it is settable to a future value if set to the current day, because Anki will still consider current day as tomorrow.

I'll fix it as you say just to use the local timezone in the UI for now and try and fix it so that the backend doesn't allow it. 

If the user is really tenacious they could still change the rollover after they change the date but they would almost certainly be doing that intentionally to break the system and it would be difficult to check every deck config when the preferences are changed.

File: rslib/src/deckconfig/update.rs [hunk 3]
```diff
@@ -613,4 +623,55 @@ mod test {
         col.set_default_relearn_steps(vec![1.]);
         assert_eq!(col.get_first_card().remaining_steps, 1);
     }
+
+    #[test]
+    fn should_clamp_ignore_revlogs_before_date_to_today() {
+        let mut col = Collection::new();
+        let today = TimestampSecs::now().date_string();
```
Comment (fernandolins) [inline rslib/src/deckconfig/update.rs]:
```suggestion
        let today = col.timing_today().unwrap().next_day_at.adding_secs(-24 * 60 * 60).date_string();
```

Luc, there might be an issue here. Initially, the test passes, especially during the day, but it might fail at night due to the time.

The idea is to adjust the "today" setting so that it passes at any time.

Comment (fernandolins) [reply rslib/src/deckconfig/update.rs]: Also, I missed the previous test: the "empty case." Could you write it again?

**Review by fernandolins — APPROVED** [review]
